server/src/main/java/org/elasticsearch/search/aggregations/timeseries/TimeSeriesIndexSearcher.java

```diff
@@ -0,0 +1,121 @@
+ * or more contributor license agreements. Licensed under the Elastic License
+ * 2.0 and the Server Side Public License, v 1; you may not use this file except
+ * in compliance with, at your election, the Elastic License 2.0 or the Server
+ * Side Public License, v 1.
+ */
+
+package org.elasticsearch.search.aggregations.timeseries;
+
+import org.apache.lucene.index.LeafReaderContext;
+import org.apache.lucene.index.SortedNumericDocValues;
+import org.apache.lucene.index.SortedSetDocValues;
+import org.apache.lucene.search.DocIdSetIterator;
+import org.apache.lucene.search.IndexSearcher;
+import org.apache.lucene.search.LeafCollector;
+import org.apache.lucene.search.Query;
+import org.apache.lucene.search.Scorer;
+import org.apache.lucene.search.Weight;
+import org.apache.lucene.util.Bits;
+import org.apache.lucene.util.BytesRef;
+import org.apache.lucene.util.PriorityQueue;
+import org.elasticsearch.cluster.metadata.DataStream;
+import org.elasticsearch.index.mapper.TimeSeriesIdFieldMapper;
+import org.elasticsearch.search.aggregations.BucketCollector;
+import org.elasticsearch.search.aggregations.LeafBucketCollector;
+
+import java.io.IOException;
+
+/**
+ * An IndexSearcher wrapper that executes the searches in time-series indices by traversing them by tsid and timestamp
+ * TODO: Convert it to use index sort instead of hard-coded tsid and timestamp values
+ */
+public class TimeSeriesIndexSearcher {
+
+    private final IndexSearcher searcher;
+
+    public TimeSeriesIndexSearcher(IndexSearcher searcher) {
+        this.searcher = searcher;
+    }
+
+    public void search(Query query, BucketCollector bucketCollector) throws IOException {
+        query = searcher.rewrite(query);
+        Weight weight = searcher.createWeight(query, bucketCollector.scoreMode(), 1);
+        PriorityQueue<LeafWalker> queue = new PriorityQueue<>(searcher.getIndexReader().leaves().size()) {
+            @Override
+            protected boolean lessThan(LeafWalker a, LeafWalker b) {
+                int res = a.tsid.compareTo(b.tsid);
+                if (res == 0) {
+                    return a.timestamp < b.timestamp;
+                } else {
+                    return res < 0;
+                }
+            }
+        };
+        for (LeafReaderContext leaf : searcher.getIndexReader().leaves()) {
+            LeafBucketCollector leafCollector = bucketCollector.getLeafCollector(leaf);
+            Scorer scorer = weight.scorer(leaf);
+            if (scorer != null) {
+                LeafWalker walker = new LeafWalker(leaf, scorer, leafCollector);
+                if (walker.next()) {
+                    queue.add(walker);
+                }
+            }
+        }
+        while (queue.top() != null) {
+            LeafWalker walker = queue.top();
+            walker.collectCurrent();
+            if (walker.next()) {
+                queue.updateTop();
+            } else {
+                queue.pop();
+            }
+        }
+    }
```

Comment:
I believe that the way we are doing search by-passes checks whether the query has been cancelled or timed out, so we'll need to add this logic to this search implementation?

Reply:
That's a good point. I hadn't thought about cancelation when I was looking last.

Reply:
Indeed, we will need to add this logic. I will do it as a follow up.